crates/kira/src/arena/iter.rs

```diff
@@ -74,7 +80,23 @@ impl<'a, T> Iterator for IterMut<'a, T> {
 
 	fn next(&mut self) -> Option<Self::Item> {
 		if let Some(index) = self.next_occupied_slot_index {
-			let slot = &mut self.arena.slots[index as usize];
+			let index_usize = usize::from(index);
+			// Since next_occupied_slot_index isn't always increasing/increasing there aren't any
+			// easy safe ways to iterate following this order.
+			let slot = if index_usize < self.slots.len() {
```

Comment:
why do we need this check if `next_occupied_slot_index` always points to a valid slot (which it does as far as i'm aware)?

Reply:
I was just being conservative and keeping the same panic on out of bounds behavior that was present with indexing.

Reply:
I can change it if you want?

Reply:
does removing the check result in a noticeable performance change?

Reply:
I did some benchmarking with the benchmarks that kira already has (after confirming that inserting delays here would impact those benchmarks).

I tried 3 different variants plus the original code before this PR.

With a bound check (i.e. what I currently have)
```rs
			let slot = if index_usize < self.slots.len() {
				let start_ptr = self.slots.cast::<ArenaSlot<T>>();
				let slot_ptr = unsafe { start_ptr.add(index_usize) };
				unsafe { slot_ptr.as_mut::<'a>() }.unwrap()
			} else {
				unreachable!("next_occupied_slot_index should always point to valid slot");
			};
```

A bound check that calls a cold method on failure, I think the panic macros theoretically already call into cold methods so this might not have any impact:
```rs
			#[cold]
			#[inline(never)]
			fn oob() -> ! {
				unreachable!("next_occupied_slot_index should always point to valid slot");
			}
			let slot = if index_usize < self.slots.len() {
				let start_ptr = self.slots.cast::<ArenaSlot<T>>();
				let slot_ptr = unsafe { start_ptr.add(index_usize) };
				unsafe { slot_ptr.as_mut::<'a>() }.unwrap()
			} else {
				oob()
			};
```

Without the check:
```rs
			let slot = {
				let start_ptr = self.slots.cast::<ArenaSlot<T>>();
				let slot_ptr = unsafe { start_ptr.add(index_usize) };
				unsafe { slot_ptr.as_mut::<'a>() }.unwrap()
			};
```

I didn't get super consistent results but it looks removing the check can be noticeable.


Here is the benchmarking data. I ran benchmarks for each case multiple times to get a feel for the variance. They are listed in the order I ran them (with the timings for specific benchmarks grouped up).

<details><summary>Benchmarks</summary>
<p>

```
no check:

simple
    time:   [4.7107 ms 4.7402 ms 4.7702 ms]
    time:   [4.6391 ms 4.6664 ms 4.6932 ms] change: [-2.4589% -1.5582% -0.7028%] (p = 0.00 < 0.05)
    time:   [4.6833 ms 4.7092 ms 4.7339 ms] change: [+0.1174% +0.9166% +1.7530%] (p = 0.02 < 0.05)
with on_start_processing callback
    time:   [4.6541 ms 4.6753 ms 4.6967 ms]
    time:   [4.6677 ms 4.6937 ms 4.7233 ms] change: [-0.3554% +0.3928% +1.1457%] (p = 0.31 > 0.05)
    time:   [4.7799 ms 4.8029 ms 4.8270 ms] change: [+1.5208% +2.3265% +3.1461%] (p = 0.00 < 0.05)

---------------------------------
bound check:

simple                  
    time:   [4.8164 ms 4.8297 ms 4.8442 ms] change: [+1.9508% +2.5596% +3.1945%] (p = 0.00 < 0.05)
    time:   [4.8281 ms 4.8472 ms 4.8677 ms] change: [-0.1694% +0.3626% +0.8488%] (p = 0.17 > 0.05)
    time:   [4.8830 ms 4.9051 ms 4.9293 ms] change: [+0.5897% +1.1947% +1.8129%] (p = 0.00 < 0.05)
with on_start_processing callback
    time:   [5.1012 ms 5.1118 ms 5.1223 ms] change: [+5.8684% +6.4306% +6.9852%] (p = 0.00 < 0.05)
    time:   [5.1057 ms 5.1164 ms 5.1270 ms] change: [-0.1761% +0.0900% +0.4053%] (p = 0.55 > 0.05)
    time:   [5.0938 ms 5.1050 ms 5.1159 ms] change: [-0.5059% -0.2224% +0.0751%] (p = 0.15 > 0.05)
---------------------------------
original:

simple
    time:   [4.7240 ms 4.7354 ms 4.7473 ms] change: [-3.9613% -3.4592% -2.9521%] (p = 0.00 < 0.05)
    time:   [4.7592 ms 4.7711 ms 4.7833 ms] change: [+0.4095% +0.7542% +1.1201%] (p = 0.00 < 0.05)
    time:   [4.8279 ms 4.8423 ms 4.8573 ms] change: [+1.0894% +1.4904% +1.8730%] (p = 0.00 < 0.05)
with on_start_processing callback
    time:   [5.0619 ms 5.0734 ms 5.0851 ms] change: [-0.9451% -0.6186% -0.3069%] (p = 0.00 < 0.05)
    time:   [5.1284 ms 5.1425 ms 5.1568 ms] change: [+0.9949% +1.3620% +1.7591%] (p = 0.00 < 0.05)
    time:   [5.0289 ms 5.0380 ms 5.0474 ms] change: [-2.3604% -2.0322% -1.7038%] (p = 0.00 < 0.05)
---------------------------------
bound check with cold method:

simple
    time:   [4.6790 ms 4.7050 ms 4.7307 ms] change: [-3.4834% -2.8341% -2.1941%] (p = 0.00 < 0.05)
    time:   [4.7144 ms 4.7347 ms 4.7546 ms] change: [-0.0581% +0.6304% +1.3303%] (p = 0.08 > 0.05)
    time:   [4.7430 ms 4.7649 ms 4.7874 ms] change: [+0.0597% +0.6391% +1.3167%] (p = 0.05 > 0.05)
with on_start_processing callback
    time:   [4.8184 ms 4.8383 ms 4.8592 ms] change: [-4.4206% -3.9643% -3.4573%] (p = 0.00 < 0.05)
    time:   [4.8480 ms 4.8660 ms 4.8851 ms] change: [-0.0065% +0.5731% +1.1490%] (p = 0.05 > 0.05)
    time:   [4.7732 ms 4.7891 ms 4.8063 ms] change: [-2.0590% -1.5793% -1.0478%] (p = 0.00 < 0.05)
```
</p>
</details> 

These are with some adjustments to try to disable CPU frequency boosting for more consistent results but I didn't really see much increase in consistency:

<details><summary>Benchmarks</summary>
<p>

```
bound check with cold method:

simple
    time:   [5.0881 ms 5.1097 ms 5.1321 ms]
    time:   [4.9674 ms 4.9978 ms 5.0285 ms] change: [-2.8775% -2.1900% -1.4783%] (p = 0.00 < 0.05)
with on_start_processing callback
    time:   [5.2004 ms 5.2385 ms 5.2781 ms] change: [-1.0833% -0.1328% +0.8674%] (p = 0.79 > 0.05)
    time:   [5.1345 ms 5.1638 ms 5.1922 ms] change: [-2.3483% -1.4264% -0.4830%] (p = 0.00 < 0.05)
---------------------------------
bound check:

simple
    time:   [5.0899 ms 5.1144 ms 5.1394 ms] change: [+1.5371% +2.3337% +3.1073%] (p = 0.00 < 0.05)
    time:   [5.2335 ms 5.2462 ms 5.2589 ms] change: [+2.0161% +2.5778% +3.1321%] (p = 0.00 < 0.05)
with on_start_processing callback
    time:   [5.0973 ms 5.1197 ms 5.1444 ms] change: [-1.5836% -0.8525% -0.1209%] (p = 0.02 < 0.05)
    time:   [5.5378 ms 5.5600 ms 5.5869 ms] change: [+7.9051% +8.5986% +9.3169%] (p = 0.00 < 0.05)
---------------------------------
no check:

simple
    time:   [5.0564 ms 5.0814 ms 5.1050 ms] change: [-3.6717% -3.1411% -2.6228%] (p = 0.00 < 0.05)
    time:   [5.1885 ms 5.2053 ms 5.2289 ms] change: [+1.8364% +2.4369% +3.0673%] (p = 0.00 < 0.05)
    time:   [5.2485 ms 5.2595 ms 5.2710 ms] change: [+0.5540% +1.0417% +1.4487%] (p = 0.00 < 0.05)
with on_start_processing callback
    time:   [5.0467 ms 5.0723 ms 5.1000 ms] change: [-9.3468% -8.7706% -8.1340%] (p = 0.00 < 0.05)
    time:   [5.4289 ms 5.4471 ms 5.4670 ms] change: [+6.6949% +7.3892% +8.0699%] (p = 0.00 < 0.05)
    time:   [5.5480 ms 5.5610 ms 5.5745 ms] change: [+1.6366% +2.0898% +2.5147%] (p = 0.00 < 0.05)
---------------------------------
bound check with cold method:

simple
    time:   [5.0197 ms 5.0434 ms 5.0673 ms] change: [-4.5992% -4.1078% -3.6129%] (p = 0.00 < 0.05)
    time:   [5.0675 ms 5.0883 ms 5.1095 ms] change: [+0.2392% +0.8902% +1.5046%] (p = 0.01 < 0.05)
with on_start_processing callback
    time:   [5.1181 ms 5.1389 ms 5.1612 ms] change: [-8.0132% -7.5901% -7.1474%] (p = 0.00 < 0.05)
    time:   [5.2394 ms 5.2623 ms 5.2869 ms] change: [+1.7712% +2.4016% +3.0434%] (p = 0.00 < 0.05)
---------------------------------
bound check:

simple
    time:   [4.9721 ms 5.0071 ms 5.0398 ms] change: [-2.3491% -1.5966% -0.8671%] (p = 0.00 < 0.05)
    time:   [5.0520 ms 5.0777 ms 5.1034 ms] change: [+0.5659% +1.4102% +2.2734%] (p = 0.00 < 0.05)
with on_start_processing callback
    time:   [4.8991 ms 4.9479 ms 4.9978 ms] change: [-6.9676% -5.9739% -4.9378%] (p = 0.00 < 0.05)
    time:   [5.0789 ms 5.0937 ms 5.1106 ms] change: [+1.8748% +2.9453% +4.0367%] (p = 0.00 < 0.05)
---------------------------------
no check:

simple
    time:   [4.6310 ms 4.6621 ms 4.6921 ms] change: [-8.8952% -8.1853% -7.4097%] (p = 0.00 < 0.05)
    time:   [4.6918 ms 4.7214 ms 4.7499 ms] change: [+0.4059% +1.2718% +2.2068%] (p = 0.01 < 0.05)
with on_start_processing callback
    time:   [4.7465 ms 4.7743 ms 4.8044 ms] change: [-6.8701% -6.2706% -5.6364%] (p = 0.00 < 0.05)
    time:   [4.8191 ms 4.8440 ms 4.8713 ms] change: [+0.6113% +1.4598% +2.2259%] (p = 0.00 < 0.05)
---------------------------------
original:

simple
    time:   [4.9678 ms 5.0116 ms 5.0539 ms] change: [+4.9959% +6.1479% +7.3438%] (p = 0.00 < 0.05)
    time:   [5.0012 ms 5.0362 ms 5.0688 ms] change: [-0.6158% +0.4893% +1.6086%] (p = 0.38 > 0.05)
with on_start_processing callback
    time:   [5.0634 ms 5.0915 ms 5.1214 ms] change: [+4.3072% +5.1099% +5.9255%] (p = 0.00 < 0.05)
    time:   [5.0659 ms 5.0938 ms 5.1234 ms] change: [-0.7835% +0.0461% +0.8308%] (p = 0.91 > 0.05)
```

</p>
</details>

Reply:
It seems reasonable to remove if you are confident that it is always in bounds.